auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/RefreshUserInfoController.kt

```diff
@@ -41,10 +42,50 @@ class RefreshUserInfoController(
             ).map(user.memberOfCNs)
 
             logger.info("Retrieved updated user info")
-            call.respond(UserInfoResponse(user, samlToken.token, roles, samlToken.expiry.epochSecond, session.isSso))
+            UserInfoResponse(user, samlToken.token, roles, samlToken.expiry.epochSecond, session.isSso)
         }
     }
 
+    suspend fun refreshUserInfo(call: ApplicationCall) {
+        val session = call.principal<OAuthSession>()!!
+        ensureNotAlreadyImpersonating(session)
+        val user = userLookupService.lookupUserByCn(session.userCn)
+        call.respond(getUserInfo(call, user))
+    }
+
+    suspend fun impersonateUser(call: ApplicationCall) {
+        val session = call.principal<OAuthSession>()!!
+        ensureNotAlreadyImpersonating(session)
+        val impersonatedUsersCn = Strings.nullToEmpty(call.parameters["userToImpersonate"]).replace("@", "!")
+        val originalUser = userLookupService.lookupUserByCn(session.userCn)
+        if (!originalUser.memberOfCNs.contains(DeltaConfig.DATAMART_DELTA_ADMIN) || !originalUser.accountEnabled) {
+            logger.atWarn().log("User does not have the necessary permissions to impersonate this user")
+            throw ApiError(
+                HttpStatusCode.Forbidden,
+                "forbidden",
+                "User is not an enabled admin",
+                "You do not have the necessary permissions to do this"
+            )
+        }
+        val userToImpersonate = userLookupService.lookupUserByCn(impersonatedUsersCn)
+        val originalUserWithImpersonatedRoles = originalUser.copy(
+            memberOfCNs = userToImpersonate.memberOfCNs,
+            firstName = "Impersonating " + userToImpersonate.firstName,
+            lastName = userToImpersonate.lastName
```

Comment:
Worth checking first name and last name don't get saved in any audit logs or anything by Delta. Or we could change the user display in Delta instead of passing a different name in here - we'd only have the email address (cn), but that might be neater/less risky.

Reply:
Changed it to use the cn/email address instead. Didn't seem to be saved to the audit logs but didn't check to thoroughly as it's still probably better to use the cn